arangod/Cluster/ActionBase.cpp

```diff
@@ -0,0 +1,261 @@
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Kaveh Vahedipour
+/// @author Matthew Von-Maszewski
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "Cluster/ActionBase.h"
+#include "Cluster/ClusterFeature.h"
+#include "Cluster/MaintenanceFeature.h"
+
+using namespace arangodb;
+using namespace arangodb::application_features;
+using namespace arangodb::maintenance;
+
+const char ActionBase::KEY[]="key";
+const char ActionBase::FIELDS[]="fields";
+const char ActionBase::TYPE[]="type";
+const char ActionBase::INDEXES[]="indexes";
+const char ActionBase::INDEX[]="index";
+const char ActionBase::SHARDS[]="shards";
+const char ActionBase::DATABASE[]="database";
+const char ActionBase::COLLECTION[]="collection";
+const char ActionBase::EDGE[]="edge";
+const char ActionBase::NAME[]="name";
+const char ActionBase::ID[]="id";
+const char ActionBase::LEADER[]="theLeader";
+const char ActionBase::LOCAL_LEADER[]="localLeader";
+const char ActionBase::GLOB_UID[]="globallyUniqueId";
+const char ActionBase::OBJECT_ID[]="objectId";
+
+inline static uint64_t secs_since_epoch() {
+  return uint64_t(
+    std::chrono::duration_cast<std::chrono::seconds>(
+      std::chrono::system_clock::now().time_since_epoch()).count());
+}
+
+ActionBase::ActionBase(MaintenanceFeature& feature, ActionDescription const& desc)
+  : _feature(feature), _description(desc), _state(READY),
+    _actionCreated(secs_since_epoch()), _progress(0) {
+  
+  _hash = _description.hash();
+  _clientId = std::to_string(_hash);
+  _id = _feature.nextActionId();
+  
+}
+
+ActionBase::ActionBase(MaintenanceFeature& feature,  ActionDescription&& desc)
+  : _feature(feature), _description(std::move(desc)), _state(READY),
+    _actionCreated(secs_since_epoch()), _progress(0) {
+  
+  _hash = _description.hash();
+  _clientId = std::to_string(_hash);
+  _id = _feature.nextActionId();
+  
+}
+
+ActionBase::~ActionBase() {
+}
+
+
+bool ActionBase::first() {
+  _actionStarted = secs_since_epoch();
+  return false;
+}
+
+void ActionBase::complete() {
+  _actionDone = secs_since_epoch();
+  auto cf = ApplicationServer::getFeature<ClusterFeature>("Cluster");
+  if (cf != nullptr) {
+    cf->syncDBServerStatusQuo();
+  }
+  _state = COMPLETE;
+}
+
+void ActionBase::fail() {
+  _actionDone = secs_since_epoch();
+  auto cf = ApplicationServer::getFeature<ClusterFeature>("Cluster");
+  if (cf != nullptr) {
+    cf->syncDBServerStatusQuo();
```

Comment:
Is there any chance that syncDBServerStatusQuo() will indirectly cause the execution of something that might want to know the state of this action.  If so, should this code and ActionBase::complete() set _state before calling syncDBServerStatusQuo()?

Reply:
no the state of this action is of no importance

Reply:
the schmutz runs only once at same time and only notes down that next run is needed.